core/codenvy-hosted-machine-authentication/src/main/java/com/codenvy/machine/authentication/server/AuthLinksInjector.java

```diff
@@ -0,0 +1,138 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.ws.rs.HttpMethod;
+import javax.ws.rs.core.UriBuilder;
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collection;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static org.eclipse.che.api.core.model.workspace.WorkspaceStatus.RUNNING;
+import static org.eclipse.che.api.core.util.LinksHelper.createLink;
+import static org.eclipse.che.api.machine.shared.Constants.TERMINAL_REFERENCE;
+import static org.eclipse.che.api.machine.shared.Constants.WSAGENT_REFERENCE;
+import static org.eclipse.che.api.machine.shared.Constants.WSAGENT_WEBSOCKET_REFERENCE;
+import static org.eclipse.che.api.workspace.shared.Constants.LINK_REL_STOP_WORKSPACE;
+
+/**
+ * Helps to inject {@link WorkspaceService} related links.
+ *
+ * @author Anton Korneta
+ */
+public class AuthLinksInjector extends WorkspaceServiceLinksInjector {
+    private static final Logger LOG           = LoggerFactory.getLogger(AuthLinksInjector.class);
+    private static final String MACHINE_TOKEN = "token";
+
+    private final String                 tokenServiceBaseUrl;
+    private final HttpJsonRequestFactory httpJsonRequestFactory;
+
+    @Inject
+    public AuthLinksInjector(@Named("che.ide.context") String ideContext,
+                             @Named("api.endpoint") String apiEndpoint,
+                             HttpJsonRequestFactory httpJsonRequestFactory) {
+        super(ideContext);
+        this.tokenServiceBaseUrl = apiEndpoint + "/machine/token/";
+        this.httpJsonRequestFactory = httpJsonRequestFactory;
+    }
+
+    protected void injectRuntimeLinks(WorkspaceDto workspace, URI ideUri, UriBuilder uriBuilder) {
+        final WorkspaceRuntimeDto runtime = workspace.getRuntime();
+        // add links for running workspace
+        if (workspace.getStatus() == RUNNING && runtime != null) {
+            runtime.getLinks()
+                   .add(createLink("DELETE",
+                                   uriBuilder.clone()
+                                             .path(WorkspaceService.class, "stop")
+                                             .build(workspace.getId())
+                                             .toString(),
+                                   LINK_REL_STOP_WORKSPACE));
+            String token = null;
+            try {
+                token = httpJsonRequestFactory.fromUrl(tokenServiceBaseUrl + workspace.getId())
+                                              .setMethod(HttpMethod.GET)
+                                              .request()
+                                              .asDto(MachineTokenDto.class).getMachineToken();
+            } catch (ApiException | IOException ex) {
+                LOG.warn("Failed to get machine token", ex);
+            }
+            final String machineToken = firstNonNull(token, "");
```

Comment:
We can't just remove the token variable because we need to get effective final variable, but we can assign it to empty string in catch block(I prefer the current version).